File: src/main/java/slavsquatsuperstar/mayonez/assets/Assets.java
package slavsquatsuperstar.mayonez.assets;

import slavsquatsuperstar.mayonez.Logger;

import java.io.File;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;

/**
 * A utility class for managing the application's resources.
 *
 * @author SlavSquatSuperstar
 */
public final class Assets {

    private static final HashMap<String, Asset> ASSETS = new HashMap<>();

    private Assets() {}

    // Asset Getters and Setters

    public static boolean hasAsset(String filename) {
        return ASSETS.containsKey(filename);
    }

    /**
     * Creates a new asset and stores it for future use.
     *
     * @param filename    The location of the sprite inside "assets/".
     * @param isClasspath Whether this asset is a classpath resource and not an external file
     */
    public static void createAsset(String filename, boolean isClasspath) {
        if (!hasAsset(filename)) {
            ASSETS.put(filename, new Asset(filename, isClasspath));
            Logger.log("Created asset \"%s\"", filename);
        } else {
            Logger.log("Asset \"%s\" already exists", filename);
        }
    }

    public static Asset getAsset(String filename, boolean isClassPath) {
        if (!hasAsset(filename))
            createAsset(filename, isClassPath);
        return ASSETS.get(filename);
    }

    // File URL Methods

    /**
     * Accesses a classpath file from within the .jar executable.
     *
     * @param path The location of the file inside the root resources directory.
     * @return The file's URL.
     */
    static URL getClasspathURL(String path) {
        return ClassLoader.getSystemResource(path);
    }

    /**
     * Accesses an external file from outside the .jar executable.
     *
     * @param filename The location of the file inside the computer's local storage.
     * @return The file's URL.
     */
    static URL getFileURL(String filename) {
        try {
            return new File(filename).toURI().toURL();
        } catch (MalformedURLException e) {
            Logger.log("Assets: Invalid file path \"%s\"", filename);
        }
        return null;
    }

}

File: src/main/kotlin/slavsquatsuperstar/mayonez/fileio/Assets.kt
package slavsquatsuperstar.mayonez.fileio

import slavsquatsuperstar.mayonez.Logger.log
import java.util.HashMap
import java.io.File
import java.net.MalformedURLException
import java.net.URL

/**
 * A utility class for managing the application's resources.
 *
 * @author SlavSquatSuperstar
 */
object Assets {

    private val ASSETS = HashMap<String, Asset>()

    // Asset Getters and Setters

    @JvmStatic
    fun hasAsset(filename: String): Boolean {
        return ASSETS.containsKey(filename)
    }

    /**
     * Creates a new asset and stores it for future use.
     *
     * @param filename    The location of the sprite inside "assets/".
     * @param isClasspath Whether this asset is a classpath resource and not an external file
     */
    @JvmStatic
    fun createAsset(filename: String, isClasspath: Boolean) {
        if (!hasAsset(filename)) {
            ASSETS[filename] = Asset(filename, isClasspath)
            log("Created asset \"%s\"", filename)
        } else {
            log("Asset \"%s\" already exists", filename)
        }
    }

    @JvmStatic
    fun getAsset(filename: String, isClassPath: Boolean): Asset? {
        if (!hasAsset(filename)) createAsset(filename, isClassPath)
        return ASSETS[filename]
    }

    // File URL Methods

    /**
     * Accesses a classpath file from within the .jar executable.
     *
     * @param path The location of the file inside the root resources directory.
     * @return The file's URL.
     */
    @JvmStatic
    fun getClasspathURL(path: String?): URL? {
        return ClassLoader.getSystemResource(path)
    }

    /**
     * Accesses an external file from outside the .jar executable.
     *
     * @param filename The location of the file inside the computer's local storage.
     * @return The file's URL.
     */
    @JvmStatic
    fun getFileURL(filename: String?): URL? {
        try {
            return File(filename!!).toURI().toURL()
        } catch (e: MalformedURLException) {
            log("Assets: Invalid file path \"%s\"", filename)
        }
        return null
    }
}